force_bdss/mco/optimizer_engines/aposteriori_optimizer_engine.py

```diff
@@ -29,19 +29,23 @@ class AposterioriOptimizerEngine(BaseOptimizerEngine):
     #: callable
     optimizer = Instance(IOptimizer, transient=True)
 
-    def optimize(self, *vargs):
+    def optimize(self, **kwargs):
         """ Generates optimization results.
 
         Yields
         ----------
-        optimization result: tuple(np.array, np.array, list)
-            Point of evaluation, objective value
+        optimization result: tuple(array-like, array-like)
+            MCO parameter and KPI values at point of optimization
         """
+        self._kpi_cache = {}
+
         #: get pareto set
         for point in self.optimizer.optimize_function(
                 self._score,
-                self.parameters):
-            kpis = self._score(point)
+                self.parameters,
+                **kwargs):
+            # Retrieve the cached raw KPI values
+            kpis = self._kpi_cache[tuple(point)]
             yield point, kpis
 
```

Comment:
So you don't have to worry about self._kpi_cache - its just updates on every _score() which is called by the optimizer?

Reply:
Yes, you can just refer it during the `optimize` method like here. This way all the BDSS framework objects can be considered in the `BaseOptimizerEngine` subclasses and the `IOptimizer` classes just deal with objects from the wrapper library (i.e. nevergrad)